dndoptions.sty

```diff
@@ -0,0 +1,83 @@
+\ExplSyntaxOn
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% Class / Package options
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+% Provides class / package option processing in of l3keys in LaTeX2e
+\RequirePackage {l3keys2e}
+
+% deprecate warning used in options and throughout the collection
+\msg_new:nnn {dnd} {deprecated} { #1 ~ is ~ deprecated ~ and ~ will ~ be ~ removed ~ in ~ version ~ #2. ~ #3 }
+
+\cs_new_protected:Npn { \__dnd_deprecated:nnn } #1#2#3
+  {
+    \msg_warning:nnnnn { dnd } { deprecated } {#1} {#2} {#3}
+  }
+
+\cs_new_protected:Npn { \__dnd_deprecated:nn } #1#2
+  {
+    \msg_warning:nnnnn { dnd } { deprecated } {#1} {#2} {}
+  }
+
+% These booleans are not set directly by an option (i.e. instead set by .code)
+\bool_new:N \l__dnd_show_background_bool
+\bool_new:N \l__dnd_show_footer_scroll_bool
+\bool_new:N \l__dnd_justified_bool
+\bool_new:N \l__dnd_multitoc_bool
+\bool_new:N \l__dnd_layout_bool
+
+\bool_set_true:N \l__dnd_multitoc_bool
+\bool_set_eq:NN \l__dnd_layout_bool \c__dnd_isclass_bool
+
+\keys_define:nn { dnd / options }
+  {
+    bg .choice:,
+    bg / full .code:n =
+      {
+        \bool_set_true:N \l__dnd_show_background_bool
+        \bool_set_true:N \l__dnd_show_footer_scroll_bool
+      },
+    bg / print .code:n =
+      {
+        \bool_set_true:N \l__dnd_show_footer_scroll_bool
+      },
+    bg / none .code:n = {},
+    bg .initial:n = full,
+    justified .code:n =
+      {
+        \bool_set_true:N \l__dnd_justified_bool,
+      },
+    justified .value_forbidden:n = true,
+    nomultitoc .code:n =
+      {
+        \bool_set_false:N \l__dnd_multitoc_bool
+      },
+    nomultitoc .value_forbidden:n = true,
+    layout .bool_set:N       = \l__dnd_layout_bool,
```

Comment:
The options don't seem to work. If you add them to the `documentclass` declaration, you get this error message:

```
Package: l3keys2e 2018-04-30 LaTeX2e option processing using LaTeX3 keys
)

! LaTeX Error: Missing \begin{document}.
```

If you add them to the `dnd` package itself, you get this error message:

```
! LaTeX Error: Unknown option `justified' for package `dnd'.

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...

l.23 ^^M

The option `justified' was not declared in package `dnd', perhaps you
misspelled its name. Try typing  <return>  to proceed.


! LaTeX Error: Unknown option `layout=true' for package `dnd'.

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...

l.23 ^^M

The option `layout=true' was not declared in package `dnd', perhaps you
misspelled its name. Try typing  <return>  to proceed.
```

CI would catch this, but CIrcleCI only automatically builds PRs based on the default branch. I'll create a "meta" PR for the `new-interface` branch so we get constant build feedback on merges.

Reply:
Okay, I got this sorted by it led to other questions. First, the class option borked because of a comma in the justified code block, but the package options error was more interesting. I thought I had tested this, but apparently not (or it was a different iteration).

In the package test, the package options are _not_ part of the `dnd` package; they are part of the `dndoptions` package. This is fixed by changing the line that includes the dndoptions.sty file to:

```latex
% used to be \RequirePackage {dndoptions}
\input {dndoptions.sty}
```

This makes the contents of dndoptions.sty part of the `dnd` package instead of `dndoptions`. This poses some questions:
1. Should we make this file a '.tex' file, the "normal" extension for `\input`?"
2. If we do that, should we rename all the sub-packages extensions the same way and include them with `\input`?

The sub-packages are not well segregated and use parts of each other. At this point, it would be tricky for a use to `\usepackage{dndmonster}.

I will take a look at the expl3 sources to see how they handle these questions, but I am interested to hear what others think.

Reply:
Found the following seldom-used extensions:
* .def - definition file that includes part of a .sty or .cls file.
* .clo - class options file

I also just noticed that the `\details` function is showing all items even if values are not supplied. I don't know when this showed up or whether the behavior will survive once the new monster interface is merged in, but it is something to keep an eye on.

Reply:
> Found the following seldom-used extensions…

Cool, these work well.

> The sub-packages are not well segregated and use parts of each other. At this point, it would be tricky for a use to `\usepackage{dndmonster}

That's true, but things work right now. We can figure out the best way to handle internal dependencies later.

> I also just noticed that the `\details` function is showing all items even if values are not supplied.

I used `git bisect` to find out when that changed. That bug is present in this branch, but not 0.7.0. The package options rewrite introduced it (18bca859beb5f1bbda9fb5f4cfea51293c8a1a8e).

Nothing leaps out at me, though. That commit didn't touch `\details`. I thought it might be something to do with loading `keycommand` with `\ExplSyntaxOn`, but that wasn't the case either.

Since we do away with `keycommand` in the new stat block, I don't think we need to get to the bottom of this.